### ISSUE TITLE
Add CP support

### ISSUE THREAD

**Issue by mikasenghaas** [PrimeIntellect-ai/prime-rl#739]
Currently, we cannot scale the training to long ctx (e.g. 64k) as the vram required to store activations is too large. Our only form of parallelism is FSDP which, being a DP dimension, does not allow scaling the micro batch size per gpu below 1. Supporting context parallelism (CP) and or tensor parallelism (TP) would allow us to split the computation for a sample between GPUs, allowing for micro batch size per gpu below 1.

Todo:
- [x] Add tp and cp to config
- [x] Create the correct device mesh in `model.py` given the passed config
- [x] `fully_shard` parameters correctly based on mesh
- [x] All models need to have an SDPA attention implementation to support CP (torch ring attn impl monkeypatches sdpa call)
- [ ] Loss needs to support local chunk computation 
  - [x] SFT
  - [ ] RL
- [x] Dataloader needs to support gpus loading the same data if they are part of the same CP/TP process group
- [x] Input ids need to be split by CP in the context_parallel context (is padding ok here?)
- [x] Add check for seqlen divisor
